algebra-core/src/curves/batch_verify.rs

```diff
@@ -0,0 +1,144 @@
+                        .collect::<Vec<C::Projective>>()[..],
+                    log2(num_buckets) as usize,
+                )?;
+            }
+        }
+        Ok(())
+    };
+
+    #[cfg(feature = "parallel")]
+    if num_rounds > 2 {
+        use std::sync::Arc;
+        let ref_points = Arc::new(points.to_vec());
+        // println!("Buckets: {}, Rounds: {}, security: {}, n_points: {}", num_buckets, num_rounds, security_param, points.len());
+        let mut threads = vec![];
+        for _ in 0..num_rounds {
+            let ref_points_thread = ref_points.clone();
+            threads.push(std::thread::spawn(
+                move || -> Result<(), VerificationError> {
+                    verify_points(&ref_points_thread[..])?;
+                    Ok(())
+                },
+            ));
+        }
+        for thread in threads {
+            thread.join().unwrap()?;
+        }
+    } else {
+        for _ in 0..num_rounds {
+            verify_points(points)?;
+        }
+    }
+
+    #[cfg(not(feature = "parallel"))]
+    for _ in 0..num_rounds {
+        verify_points(points)?;
+    }
+
+    Ok(())
+}
+
+pub fn batch_verify_in_subgroup_proj<C: ProjectiveCurve>(
+    points: &[C],
+    security_param: usize,
+) -> Result<(), VerificationError> {
+    let (num_buckets, num_rounds) = get_max_bucket(security_param, points.len());
+    // println!("Buckets: {}, Rounds: {}, security: {}, n_points: {}", num_buckets, num_rounds, security_param, points.len());
+    let rng = &mut thread_rng();
+
+    for _ in 0..num_rounds {
+        let mut bucket_assign = Vec::with_capacity(points.len());
+        for _ in 0..points.len() {
+            bucket_assign.push(rng.gen_range(0, num_buckets));
+        }
+        // If our batch size is too small, we do the naive bucket add
+        let zero = C::zero();
+        let mut buckets = vec![zero; num_buckets];
+        for (p, a) in points.iter().zip(bucket_assign) {
+            buckets[a].add_assign(p);
+        }
+
+        if num_buckets <= 3 {
+            if !buckets
+                .iter()
+                .all(|b| b.mul(<C::ScalarField as PrimeField>::Params::MODULUS) == C::zero())
+            {
+                return Err(VerificationError);
+            }
+        } else {
+            // println!("CALLING BUCKET PROJ RECURSIVE");
+            batch_verify_in_subgroup_proj(&buckets[..], log2(num_buckets) as usize)?;
+        }
+    }
+    Ok(())
+}
+
+// We get the greatest power of 2 number of buckets
+// such that we minimise the number of rounds
+// while satisfying the constraint that number of rounds * buckets * 2 < n
+
+// Number of buckets is always greater than new security param.
+// So only need 1 round subsequently
+fn get_max_bucket(security_param: usize, n_elems: usize) -> (usize, usize) {
+    let mut log2_num_buckets = 1;
+    let num_rounds =
+        |log2_num_buckets: usize| -> usize { (security_param - 1) / log2_num_buckets + 1 };
```

Comment:
This should be `ceil(security_param / (log2_num_buckets - 1))`, which is not the same as `security_param / (log2_num_buckets - 1) + 1` when `log2_num_buckets - 1` divides `security_param`.

Reply:
You'll need to initialize `log2_num_buckets = 2`.